### PR TITLE
CoreCLR port: Don't allocate in NullStream.CopyToAsync

### PR DESCRIPTION
**Pull request by jamesqo**
Port of dotnet/coreclr#4383 to the CoreRT version of `Stream`. Also added a couple of tests improving the coverage for `NullStream`.

List of test changes:

- Moved `NopStream` into a new file
- Added `ReadOnlyStream`, which acts like `Stream.Null` but throws when trying to write
- Added a bunch more `NullStream` tests (which are also more clean)
  - Refactored this from the previous testing scheme, which only used one method

cc @JonHanna @justinvp @stephentoub (also @jkotas since he originally requested the port)

### CODE REVIEW

File: src/System.IO/src/System/IO/Stream.cs [hunk 1]
```diff
@@ -28,6 +28,36 @@ internal SemaphoreSlim EnsureAsyncActiveSemaphoreInitialized()
             // WaitHandle, we don't need to worry about Disposing it.
             return LazyInitializer.EnsureInitialized(ref _asyncActiveSemaphore, () => new SemaphoreSlim(1, 1));
         }
+        
+        // This method is used by CopyTo, CopyToAsync,
+        // and NullStream's override of the latter.
```
Comment (justinvp) [inline src/System.IO/src/System/IO/Stream.cs]: Nit: This comment isn't in CoreCLR, is it really needed here?

File: src/System.IO/src/System/IO/Stream.cs [hunk 2]
```diff
@@ -28,6 +28,36 @@ internal SemaphoreSlim EnsureAsyncActiveSemaphoreInitialized()
...
+        internal void ValidateCopyToArguments(Stream destination, int bufferSize)
```
Comment (justinvp) [inline src/System.IO/src/System/IO/Stream.cs]: In CoreCLR this method is defined just above the nested `NullStream` class, why not put it in the same place?

File: src/System.IO/tests/Stream/Stream.NopStream.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            get { return false; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+
```
Comment (justinvp) [inline src/System.IO/tests/Stream/Stream.NopStream.cs]: Nit: There are some extra lines between members throughout this file that could be removed.

File: src/System.IO/tests/Stream/Stream.NopStream.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public override void Flush() { }
+        public override long Seek(long offset, SeekOrigin origin) { throw new NotSupportedException(); }
+        public override void SetLength(long value) { throw new NotSupportedException(); }
+        public override int Read(byte[] buffer, int offset, int count) { return 0; }
+        public override void Write(byte[] buffer, int offset, int count) { }
+
+    }
+}
```
Comment (justinvp) [inline src/System.IO/tests/Stream/Stream.NopStream.cs]: Nit: There should be a blank line at the end of the file.

File: src/System.IO/tests/Stream/Stream.NullTests.cs [hunk 5]
```diff
@@ -5,31 +5,91 @@
...
-            s.Write(new byte[2], 0, 2);
-
-            Assert.Equal(0, await s.ReadAsync(new byte[2], 0, 2));
-
-            s.Flush();
-            s.Dispose();
+            // Since Stream.Null previously inherited its CopyToAsync
+            // implementation from the base class, which did check its
+            // arguments, we have to make sure it validates them as
+            // well for compat.
```
Comment (justinvp) [inline src/System.IO/tests/Stream/Stream.NullTests.cs]: Nit: I don't have a strong preference either way, but not sure how much value this comment adds here as the Asserts below lock in the pre-existing behavior.

File: src/System.IO/tests/Stream/Stream.ReadOnly.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            get { return 0; }
+            set { }
+        }
+
+        public override void Flush() { }
+        public override long Seek(long offset, SeekOrigin origin) => 0;
+        public override void SetLength(long value) { }
+        public override int Read(byte[] buffer, int offset, int count) => 0;
+        public override void Write(byte[] buffer, int offset, int count) { throw new InvalidOperationException(); }
```
Comment (justinvp) [inline src/System.IO/tests/Stream/Stream.ReadOnly.cs]: This should throw `NotSupportException`

File: src/System.IO/tests/Stream/Stream.ReadOnly.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    }
+}
```
Comment (justinvp) [inline src/System.IO/tests/Stream/Stream.ReadOnly.cs]: Nit: There should be a blank line at the end of the file.

File: src/System.IO/tests/Stream/Stream.ReadOnly.cs (hunk 6 above)
Comment (jamesqo) [inline src/System.IO/tests/Stream/Stream.ReadOnly.cs]: Didn't know how to quite work this out, since the check this was meant to validate also throws the same exception:

```cs
if (!destination.CanWrite)  
{  
    throw new NotSupportedException(SR.NotSupported_UnwritableStream);  
}
```

So then you wouldn't be able to tell whether the check was actually working, or it was getting thrown from `ReadOnlyStream.Write`. Perhaps I should just noop here instead?

File: src/System.IO/tests/Stream/Stream.NullTests.cs (hunk 5 above)
Comment (jamesqo) [inline src/System.IO/tests/Stream/Stream.NullTests.cs]: I added it since the reader might be curious why `WriteAsync` / `ReadAsync` were being tested with invalid values, but `CopyToAsync` performed validation.

File: src/System.IO/src/System/IO/Stream.cs (hunk 1 above)
Comment (jamesqo) [inline src/System.IO/src/System/IO/Stream.cs]: Yeah, in retrospect I should probably remove this comment.

File: src/System.IO/src/System/IO/Stream.cs (hunk 2 above)
Comment (jamesqo) [inline src/System.IO/src/System/IO/Stream.cs]: I did it since the previous method is the only internal/private method in the class, and it's defined at the top of the file.

**Comment by jamesqo** [comment]
Looks like I'm getting a test failure from `UnmanagedMemoryStream` due to my changes: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/centos7.1_debug_tst_prtest/325/consoleFull#-21371504001f1a4601-6aec-4fd5-b678-78d4389fd5e8

Will fix shortly. (I didn't catch it earlier since I only ran the tests in the `System.IO` assembly, unfortunately.)

File: src/System.IO/tests/Stream/Stream.NullTests.cs [hunk 8]
```diff
@@ -5,31 +5,91 @@
...
 using System.Threading.Tasks;
 using Xunit;
 
 namespace System.IO.Tests
 {
-    public class NullCtorTests
+    public class NullTests
     {
+        [Theory]
+        [InlineData(20)]
```
Comment (stephentoub) [inline src/System.IO/tests/Stream/Stream.NullTests.cs]: Why 20?  Isn't calling them once sufficient?

File: src/System.IO/tests/Stream/Stream.NullTests.cs [hunk 9]
```diff
@@ -5,31 +5,91 @@
...
+                Stream.Null.Dispose();
+            }
+        }
+        
+        [Theory]
+        [MemberData(nameof(NullStream_CopyToData))]
+        public async static Task TestNullStream_CopyTo(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            Stream.Null.CopyTo(destination, bufferSize);
+            await Stream.Null.CopyToAsync(destination, bufferSize, cancellationToken);
```
Comment (stephentoub) [inline src/System.IO/tests/Stream/Stream.NullTests.cs]: What are these tests verifying, other than that they don't throw an exception?  Couldn't we check, for example, that the destination's position didn't change by the copy, since nothing should have been copied?

File: src/System.IO/tests/Stream/Stream.NullTests.cs [hunk 10]
```diff
@@ -5,31 +5,91 @@
...
+        public static void TestNullStream_Dispose(int times)
+        {
+            for (int i = 0; i < times; i++)
+            {
+                Stream.Null.Dispose();
+            }
+        }
+        
+        [Theory]
+        [MemberData(nameof(NullStream_CopyToData))]
```
Comment (stephentoub) [inline src/System.IO/tests/Stream/Stream.NullTests.cs]: This doesn't need to be a Theory; I'd suggest just using a MemoryStream in the body.

File: src/System.IO/tests/Stream/Stream.NullTests.cs [hunk 11]
```diff
@@ -5,31 +5,91 @@
...
+        [Theory]
+        [MemberData(nameof(NullStream_ReadWriteData))]
+        public async static Task TestNullStream_Write(byte[] buffer, int offset, int count)
+        {
+            Stream.Null.Write(buffer, offset, count);
+            await Stream.Null.WriteAsync(buffer, offset, count);
+        }
+        
+        [Theory]
+        [MemberData(nameof(NullStream_WriteByteData))]
```
Comment (stephentoub) [inline src/System.IO/tests/Stream/Stream.NullTests.cs]: There's no reason to run this test 256 times, once for each byte.  The test isn't even checking the effect of the write (which is a nop), so this can just be a Fact that writes a single byte, doesn't matter which.

File: src/System.IO/tests/Stream/Stream.ReadOnly.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+using System;
+using System.IO;
+using Xunit;
+
+namespace System.IO.Tests
+{
+    // This class emulates the behavior of Stream.Null,
+    // except CanWrite == false and attempting to write
+    // will throw a InvalidOperationException.
+    public sealed class ReadOnlyStream : Stream
```
Comment (stephentoub) [inline src/System.IO/tests/Stream/Stream.ReadOnly.cs]: Do you need to create a new stream type?  I only see it being used once, in a test that just wants to see if an exception from its WriteAsync propagates out of a CopyToAsync call, and for that you could use any Stream that throws from its Write method, such as a ```new MemoryStream(new byte[1], writable: false)```.

File: src/System.IO/tests/Stream/Stream.NopStream.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using Xunit;
+
+namespace System.IO.Tests
+{
+    public sealed class NopStream : Stream
```
Comment (stephentoub) [inline src/System.IO/tests/Stream/Stream.NopStream.cs]: Why is this necessary?  Why can't you just use an existing stream like MemoryStream?

File: src/System.IO/src/System/IO/Stream.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.IO/src/System/IO/Stream.cs]: I'd really like to keep the code as similar as possible between the two repos; I realize they've already diverged, but I'd like to avoid further divergence as much as possible.

File: src/System.IO/tests/Stream/Stream.NopStream.cs (hunk 13 above)
Comment (jamesqo) [inline src/System.IO/tests/Stream/Stream.NopStream.cs]: @stephentoub There is existing code that uses this class [here](https://github.com/dotnet/corefx/blob/master/src/System.IO/tests/Stream/Stream.TimeoutTests.cs#L54) I simply moved it to a new file. Should I move it back and delete the lines using it? (It looks like the tests in that file are already using `MemoryStream`, BTW; they just added an additional class for the sake of further testing.)

File: src/System.IO/tests/Stream/Stream.ReadOnly.cs (hunk 12 above)
Comment (jamesqo) [inline src/System.IO/tests/Stream/Stream.ReadOnly.cs]: @stephentoub Good point; I'll switch to using the `MemoryStream` constructor that makes it a read-only stream.

File: src/System.IO/tests/Stream/Stream.NullTests.cs (hunk 8 above)
Comment (jamesqo) [inline src/System.IO/tests/Stream/Stream.NullTests.cs]: Did that so the tests could verify that calling `Dispose` wouldn't have any side effects; e.g. calling `Dispose` twice wouldn't throw on the second time. In hindsight maybe making it a `Theory` and adding a parameter for the number of iterations was a bit overkill; I'll simply copy-paste the method call and add a note explaining why it's being repeated.

**Comment by jamesqo** [comment]
PR feedback has been addressed; the changes are now limited to just 2 files.

**Comment by stephentoub** [comment]
Test Innerloop Ubuntu14.04 Release Build and Test please

**Comment by stephentoub** [comment]
LGTM.  Thanks for the port and additional tests.
